geoutils/georaster/raster.py

```diff
@@ -195,31 +210,97 @@ def __init__(
 
         :return: A Raster object
         """
+        self.driver: str | None = None
+        self.name: str | None = None
+        self.nodata: int | float | list[int] | list[float] | None = None
+        self.filename: str | None = None
+        self.tags: dict[str, Any] = {}
+
+        self._data: np.ndarray | np.ma.masked_array | None = None
+        self._bands = bands
+        self._masked = masked
+        self._disk_hash: int | None = None
+        self._is_modified = True
+        self._disk_shape: tuple[int, int, int] | None = None
+        self._disk_indexes: tuple[int] | None = None
+        self._disk_dtypes: tuple[str] | None = None
+
+        # This is for Raster.from_array to work.
+        if isinstance(filename_or_dataset, dict):
+            self.data = filename_or_dataset["data"]
+            self.transform: rio.transform.Affine = filename_or_dataset["transform"]
+            self.crs: rio.crs.CRS = filename_or_dataset["crs"]
+            self.nodata = filename_or_dataset["nodata"]
+            for key in filename_or_dataset:
+                if key in ["data", "transform", "crs", "nodata"]:
+                    continue
+                setattr(self, key, filename_or_dataset[key])
+            return
+
         # If Raster is passed, simply point back to Raster
         if isinstance(filename_or_dataset, Raster):
             for key in filename_or_dataset.__dict__:
                 setattr(self, key, filename_or_dataset.__dict__[key])
             return
         # Image is a file on disk.
-        elif isinstance(filename_or_dataset, str):
-            # Save the absolute on-disk filename
-            self.filename = os.path.abspath(filename_or_dataset)
-            if as_memfile:
-                # open the file in memory
-                memfile = MemoryFile(open(filename_or_dataset, "rb"))
-                # Read the file as a rasterio dataset
-                self.ds = memfile.open()
+        elif isinstance(filename_or_dataset, (str, rio.io.DatasetReader, rio.io.MemoryFile)):
+
+            with ExitStack():
```

Comment:
Here a comment to shortly describe the behaviour of ExitStack() would help!

Reply:
+ 1

Reply:
Oh no, then I have to remember why it's there in the first place!!

Reply:
Described in b059bfb